### ISSUE TITLE
remote state partial config backend-config file triggers "migration" every time

### ISSUE THREAD

**Issue by sysadmiral** [hashicorp/terraform#13212]
### Terraform Version
```shell
$ terraform version
Terraform v0.9.2
```

### Affected Resource(s)
- s3 backend
- `terraform init`

### Terraform Configuration Files
backend_config
```shell
profile = "bar"
region  = "eu-west-1"
bucket  = "foo"
key     = "test"
encrypt = "true"
```
terraform.tf
```hcl
terraform {
  backend "s3" {
    profile    = "bar"
    region     = "eu-west-1"
    bucket     = "foo"
    key        = "test"
    encrypt    = "true"    
  }
}
```

### Expected Behavior
Given the two methods to setup a backend above (a config file or within a terraform block) I should get identical behaviour when running `terraform init` or `terraform init -backend-config=backend_config`

### Actual Behavior
If I use the terraform.tf file then the backend is configured and subsequent `terraform init` runs see that nothing is changed and carry on happily.

If I use the backend_config file and "partial configuration" then the first init sets things up but then all subsequent runs think something has changed each time. (Just out of interest - This is a little like issues with subjects like "terraform thinks my resource has changed every time" and I wonder if the mechanism to parse the values is similar?)

This is a problem because it thinks the configuration is slightly different it always prompts about migrating the state. The state location and backend in reality haven't changed at all.

### Steps to Reproduce
In an empty terraform repo with a valid aws profile for s3 access:

* copy terraform.tf into the repo
* run `terraform init` multiple times
* delete the values in the `backend "s3" {}` block in terraform.tf (setup partial configuration)
* copy backend_config to the repo
* run `terraform init -backend-config=backend_config` multiple times

**Comment by stephenchu** [comment]
This is exactly the bug behavior I am [seeing here](https://groups.google.com/forum/#!topic/terraform-tool/fPjRT6D58mo) as well. The mentioned bug-demo repo [README.md](https://github.com/stephenchu/bug-demo/blob/master/README.md) in the thread has the command line history to demonstrate the bug.

**Comment by apparentlymart** [comment]
@jbardin this sounds suspiciously like something you were working on (or talking about) recently, but I wasn't able to quickly dig up a PR. Does this look like something you've addressed already (but not included in a release) or is this a new issue?

**Comment by jbardin** [comment]
@apparentlymart, Yes, this looks to be fixed by #13165. Thanks!
